### ISSUE TITLE
TypeGuard changes type of variable outside the if block

### ISSUE THREAD

**Issue by KotlinIsland** [microsoft/pyright#2356]
```py
from typing_extensions import TypeGuard

class A:
    def foo(self) -> None: ...

def foo(it: object) -> TypeGuard[int]: ...

a = A()
if foo(a):
    pass
a  # (variable) a: int | A
```
`a` should be `A` not `int | A`

### Edit
because `a` is now a `int | A` pyright says it's invalid to call `a.foo()`

**Comment by erictraut** [comment]
This behavior is correct. When a type guard is applied (regardless of whether it's a user-defined TypeGuard or a built-in type guard), the type is retained past the block.

Consider the following:
```python
def func(a: object):
    if isinstance(a, int):
        pass
    else:
        raise ValueError()

    reveal_type(a) # int
```

**Comment by erictraut** [comment]
Another example to consider:

```python
class A: ...
class B: ...

def func(a: object):
class A:
    x: int

class B:
    x: str

def func(a: object):
    if isinstance(a, A):
        pass
    else:
        assert isinstance(a, B)

    reveal_type(a)  # mypy reveals object, pyright reveals A | B

    print(a.x) # mypy emits error, pyright doesn't
```

**Comment by KotlinIsland** [comment]
So your telling me that:
```py
class A:
    def foo(self) -> None: ...

a = A()
if isinstance(a, int):
    pass
a.foo()
```
Is not type safe?

I understand if the else branch terminates the control flow, but that's not happening here,

**Comment by KotlinIsland** [comment]
> When a type guard is applied (regardless of whether it's a user-defined TypeGuard or a built-in type guard),
```py
from typing_extensions import TypeGuard


class A:
    def foo(self) -> None: ...

def foo(it: object) -> TypeGuard[int]: ...

a = A()
if isinstance(a, int):
    pass
a.foo() # a: <subclass of A and int> | A

if foo(a):
    pass
a.foo() # error (method) foo: Unknown | () -> None
```
that's not whats happening here.

**Comment by KotlinIsland** [comment]
Something seems whacky with type guards.
TS:
```ts
class A{
    foo() {}
}

class B {
    bar() {}
}

const a  = new A
if (a instanceof B) {
    a.bar()
}

a // const a: A
```

pyright:
```py
class A:
    def foo(): ...

class B:
    def bar(): ...

a = A()
if isinstance(a, B):
    a.bar()

reveal_type(a)  # Type of "a" is "<subclass of A and B> | A"
```

**Comment by erictraut** [comment]
> Is not type safe?

It is type safe, and it passes type checking without issue in both pyright and mypy.

> these two identical guards have different behavior.

These are not identical. One is using `isinstance` semantics, the other is using the TypeGuard rules defined in PEP 647.

> Something seems whacky with type guards.

TypeScript doesn't support multiple inheritance or nominal types, so some of the details will differ when it comes to type guards.

**Comment by KotlinIsland** [comment]
![image](https://user-images.githubusercontent.com/65446343/135406514-263d388b-99c4-4c84-8c2b-79f8255549b9.png)

**Comment by KotlinIsland** [comment]
Maybe it's my pyright version or something, I'm using Python v2021.9.1246542782 in vscode.

**Comment by KotlinIsland** [comment]
Is there a reason why pylance and mypy differ on behavior here?
```py
class A:
    def foo(self):
        ...

a = A()

if isinstance(a, int):
    a.foo()

reveal_type(a)
# mypy: note: Revealed type is "__main__.A"
# pylance: Type of "a" is "<subclass of A and int> | A"

**Comment by jakebailey** [comment]
That's interesting. In the `isinstsance` case, we have a special intersection type to represent any type that inherits from both `A` and `int` (as that's all that `a` could be in that context), but when the code flow is going backwards, it's unioning that with the `else` case (which is `A`), and ending up with that odd type.

_Technically_, though, I think that that union type is always assignable to `A` anyway, so this doesn't have much practical impact.

**Comment by erictraut** [comment]
I consider the current behavior correct. This isn't a bug, it's expected behavior. I can't speak for the mypy implementation. There are going to be differences between mypy and pyright in some cases.

**Comment by KotlinIsland** [comment]
This is showing an invalid error where mypy doesn't. I don't understand how this is expected behavior.
```py

from typing import TypeGuard


def guard(it: object) -> TypeGuard[int]:
    ...

class A:
    def foo(self) -> None: ...

a = A()

if guard(a):
    pass

a.foo()   # (variable) a: int | A
# (method) foo: Unknown | () -> None
```

**Comment by erictraut** [comment]
The type guard is "narrowing" the type to an `int`, so in the `if` portion of the `if/else` statement, it becomes an `int`. In the (implied) negative (`else`) portion of the `if/else` statement, it remains an `A`. After the `if/else` statement, it is the union of the two types: `int | A`.

This is consistent with the way type narrowing works in general in pyright. Maybe this example will help clarify.

```python
def func(a: object):
    if guard(a):
        pass
    else:
        assert isinstance(a, float)

    reveal_type(a)  # pyright: int | float, mypy: object

    print(a + 1)  # pyright: no error, mypy: error
```

Mypy emits an error for a legitimate case, but pyright does not. I think that's because mypy is performing a join operation at the end of the `if/else` whereas pyright is performing a union. A union is preferable because it retains more type details.

**Comment by KotlinIsland** [comment]
So why do `isinstance` guards behave completely different?
```py
class A:
    def foo(self) -> None: ...

a = A()
if isinstance(a, int):
    pass
a.foo()   # no error
reveal_type(a) # Type of "a" is "<subclass of A and int> | A"
```

**Comment by KotlinIsland** [comment]
Ohhh, it's because with an isinstance block the type becomes an intersection of the two types, but with a `TypeGuard` it only becomes the guarded type.
